runtime/ast/unmarshal.go

```diff
@@ -0,0 +1,901 @@
+	"math/big"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/onflow/cadence/runtime/common"
+)
+
+func jsonMarshalAndVerify(i interface{}) (b []byte, err error) {
+	if b, err = json.Marshal(i); err == nil {
+		expected := string(b)
+		roundtripped := reflect.New(reflect.TypeOf(i)).Interface()
+		if err = json.Unmarshal(b, &roundtripped); err == nil {
+			if b, err = json.Marshal(roundtripped); err == nil {
+				actual := string(b)
+				if expected != actual {
+					return nil, fmt.Errorf("un/marshal failed:\n%s\n%s\n", expected, actual)
+				}
+			}
+		}
+	}
+	return
+}
+
+//////////////////////////////////////////////////////////////////////////////
+
+var accessKindMap = map[string]Access{
+	"AccessNotSpecified":   AccessNotSpecified,
+	"AccessPrivate":        AccessPrivate,
+	"AccessContract":       AccessContract,
+	"AccessAccount":        AccessAccount,
+	"AccessPublic":         AccessPublic,
+	"AccessPublicSettable": AccessPublicSettable,
+}
+
+var conditionKindMap = map[string]ConditionKind{
+	"ConditionKindUnknown": ConditionKindUnknown,
+	"ConditionKindPre":     ConditionKindPre,
+	"ConditionKindPost":    ConditionKindPost,
+}
+
+var expressionTypeMap = map[string]func() Expression{
+	"ArrayExpression":       func() Expression { return &ArrayExpression{} },
+	"BinaryExpression":      func() Expression { return &BinaryExpression{} },
+	"BoolExpression":        func() Expression { return &BoolExpression{} },
+	"CastingExpression":     func() Expression { return &CastingExpression{} },
+	"ConditionalExpression": func() Expression { return &ConditionalExpression{} },
+	"CreateExpression":      func() Expression { return &CreateExpression{} },
+	"DestroyExpression":     func() Expression { return &DestroyExpression{} },
+	"DictionaryExpression":  func() Expression { return &DictionaryExpression{} },
+	"FixedPointExpression":  func() Expression { return &FixedPointExpression{} },
+	"ForceExpression":       func() Expression { return &ForceExpression{} },
+	"FunctionExpression":    func() Expression { return &FunctionExpression{} },
+	"IdentifierExpression":  func() Expression { return &IdentifierExpression{} },
+	"IndexExpression":       func() Expression { return &IndexExpression{} },
+	"IntegerExpression":     func() Expression { return &IntegerExpression{} },
+	"InvocationExpression":  func() Expression { return &InvocationExpression{} },
+	"MemberExpression":      func() Expression { return &MemberExpression{} },
+	"NilExpression":         func() Expression { return &NilExpression{} },
+	"PathExpression":        func() Expression { return &PathExpression{} },
+	"ReferenceExpression":   func() Expression { return &ReferenceExpression{} },
+	"StringExpression":      func() Expression { return &StringExpression{} },
+	"UnaryExpression":       func() Expression { return &UnaryExpression{} },
+}
+
+var locationKindMap = map[string]func() common.Location{
+	"AddressLocation":     func() common.Location { return &common.AddressLocation{} },
```

Comment:
What is the reason for the value being a function here? Can the value type of the map just be Expression?